### ISSUE TITLE
Add UDP broadcast support

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11556]
|<img alt="pbridger's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| pbridger reported|
|-|-|
|Trac ID|trac#454|
|Type|enhancement|
|Created|2003-12-30 05:37:43Z|
|Branch|https://github.com/twisted/twisted/tree/udp-broadcast-454-3|

Attachments:

* [udpbroadcastunittest.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/14019988a92023b21c8fbafb2b615c6ce575da38/a0676e8a005b88f92cc7110aa6a0970b48e9b827.patch) (1847 bytes) - added by hawkowl on 2013-03-29 16:30:33Z - Unit Test - UDP Broadcast over loopback
* [broadcastudptestcasev2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/14019988a92023b21c8fbafb2b615c6ce575da38/9e348035c051aac24760befd0c6d08841e8a3169.patch) (2195 bytes) - added by hawkowl on 2013-08-23 10:43:50Z - Unit Test - UDP Broadcast over loopback, v2
* [broadcastudptestcasev3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/14019988a92023b21c8fbafb2b615c6ce575da38/48c58f2e37b5ed6a2b68053829bcbb9fce023eba.patch) (1872 bytes) - added by hawkowl on 2013-08-23 11:38:33Z - Unit Test - UDP Broadcast over loopback, v3
* [udpbroadcast.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/14019988a92023b21c8fbafb2b615c6ce575da38/32495121bcf03f4e82a50f153018e26f10279d9e.patch) (1750 bytes) - added by hawkowl on 2013-08-24 08:46:58Z - Implementation of set/getBroadcastAllowed and corresponding test
* [udpbroadcast-fix.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/14019988a92023b21c8fbafb2b615c6ce575da38/706b1b55054065fb860bd4bb9e0e780a306b219a.patch) (1910 bytes) - added by hawkowl on 2013-10-20 15:42:21Z - Fixes to udpbroadcast.patch, for IOCP + TwistedChecker
* [udpbroadcast.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/140/14019988a92023b21c8fbafb2b615c6ce575da38/6bf33f1944ed957df15154d4d4af9d2b9165a787.py) (3405 bytes) - added by rwall on 2013-10-22 16:53:10Z - Send UDP broadcast messages for tracking a cluster of hosts

<details><summary>Searchable metadata</summary>

```
trac-id__454 454
type__enhancement enhancement
reporter__pbridger pbridger
priority__normal normal
milestone__ 
branch__branches_udp_broadcast_454_3 branches/udp-broadcast-454-3
branch_author__hawkowl__rwall hawkowl, rwall
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1072762663000000 1072762663000000
changetime__1470178858262812 1470178858262812
version__ 
owner__hawkowl hawkowl
cc__itamarst cc__pbridger cc__matth cc__hawkowl@... cc__jerith
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="pbridger's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|pbridger commented|
|-|-|

```
#!html
<pre>
&lt;me> How can I setsockopt?
&lt;exarkun> By violating encapsulation
&lt;exarkun> itamar will fix it soon
&lt;exarkun> Or we will stone him to death
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
My solution is to add a method to transport, getHandle(). This returns the
socket object on default reactor (might return something else on e.g. Windows
reactor), which you can then call setsockopt on or whatever. I don't want to add
a method for this because behaviour is way too OS specific to make a reasonable
abstraction for.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
No comments, guess this is closed.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set status to `reopened`|
|-|-|

My solution to this would be to use the existing reactor.listenUDP(...) method [rather than create a .listenBroadcast(...) to copy how Multicasts are done in .listenMulticast(...) ] and add a get/set method on the IUDPTransport that controls the relevant socket option (SO_BROADCAST). This solution seems to be consistent with ITCPTransport's socket option methods (eg. get/setTcpNoDelay) and would make the method to enable UDP broadcasting "self.transport.setUDPBroadcast(True)" or similar.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

To add a bit more context, we had a discussion about `ISystemHandle.getHandle` and concluded that while this does make it _possible_ to set the necessary socket option for broadcast, it doesn't make it _easy_.  And since there is presumably just a correct way to do this on POSIX and a correct way to do it on Windows (are the ways even different?) having implementations of those ways under a portable interface in Twisted seems like a good idea (that's the idea of big chunks of Twisted, after all).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

Attached a unit test into test_udp.py that uses the ```self.transport.socket.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)``` method of enabling broadcast, then broadcasts a UNIX timestamp and passes if it receives the same timestamp back. If it works, then we could implement unit tests that broadcast to lo for testing whether a potential .enableBroadcast() transport method actually works.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

For context: the unit test patch is for running on buildbot so that the behaviour of the test can be seen for each of the supported platforms, to make sure that this way of enabling UDP broadcast works universally before a proper .enableBroadcast() feature is implemented.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [38112]) Branching to 'udp-broadcast-454'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [38113]) Apply udpbroadcastunittest.patch 

refs [#11556](https://github.com/twisted/twisted/issues/11556)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [38114]) Mostly trivial cleanups, but also be sure to return the Deferred from the test method so trial runs the reactor long enough for something to happen.

refs [#11556](https://github.com/twisted/twisted/issues/11556)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @hawkowl|
|-|-|

Thanks for your work on this hawkowl.  Sorry about the long review latency.

I checked your patch in with some changes (visible in the comment above) and [forced a build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/udp-broadcast-454) on buildbot to see what happens.

Some notes for future patches:

  1. Since there's a branch, now, it's probably a good idea to make patches against that instead of trunk.  This isn't strictly required, and since the code in the branch now is mostly for testing what happens on various platforms, it might make sense to just start fresh once we've learned what the results are.  Either way, just make sure it's clear whether the next patch is based on trunk or on the branch.
  1. I'd strongly consider putting the tests in `twisted/internet/test/test_udp.py` instead, and writing them in the style demonstrated by the tests already in that module.  `twisted/test/test_udp.py` is full of tests written in an older style that's not as useful (primarily, it can only test the UDP implementation of a single reactor per test suite run; the style used in `twisted/internet/test/test_udp.py` can test all of the reactors in one run).

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl commented|
|-|-|

Hi, I've come back to fix this :)

I've moved the tests to where exarkun noted they should go, incorporated the fixes that were put in before, and hopefully it matches the style guide for now.

This patch is based off current trunk (just in case 13.0/13.1 changed anything), and has the same goal as the last patch, to make sure that this is a proper method of implementing the test (or if we can test it at all), and what platforms support it.

On my machine (2.7.5, Linux64), it passes on all the reactors that run tests, so hopefully buildbot gives us all green for the rest :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl commented|
|-|-|

Jerith pointed out this patch was actually broken, since I totally didn't see that you had to start your own reactor. Also, he helped me fix up the style/where to put things, since it was pretty atrocious. New patch follows.

**Comment by twisted-trac** [comment]
|[<img alt="jerith's avatar" src="https://avatars.githubusercontent.com/u/43731?s=50" width="50" height="50">](https://github.com/jerith)<a name="note_17"></a>|@jerith commented|
|-|-|

FWIW, this test fails for me on Mac OS X 10.8 with "```socket.error: [Errno 49] Can't assign requested address```".

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_18"></a>|@hawkowl removed owner|
|-|-|

I believe that's the problem we had with 10.8 with the original tests... It works here, so I'm not sure if that's a OS X thing. It also seemed to work on 10.6, with the original patch. Hmm.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl commented|
|-|-|

Okay, so, a few things learned:
a) you can't UDP broadcast to lo
b) it silently works on Linux but hard fails on OSX and Windows (as it should)
c) I shouldn't be testing this anyway, as it's doing things that's out of the context of this patch.

I'll write the feature and the unit test and post it up here when I think it's reviewable.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_20"></a>|@hawkowl commented|
|-|-|

I've written the functionality for this ticket how I mentioned in my first post - wrapping something around setsockopt() to enable it. I set both functions to use/return a bool, breaking convention with get/setTcpNoDelay in tcp.py (which passes a raw 1/0 to setsockopt(), but returns a bool via operator.truth() in the get function), but I think it's much more 'expected'. There's also a test in there to check that the setting is applied.

I'm not 100% sure on the naming, or if the bool -> int stuff I do in setBroadcastAllowed() is messy, so feedback on that would be appreciated. I'm also guessing that documentation in the UDP section would be required, so I'll follow up with that if this gets through.

I have no idea how this will react on IOCP, so I guess we'll have to see what Buildbot says :)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_21"></a>|@wallrj commented|
|-|-|

(In [40405]) Branching to 'udp-broadcast-454-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_22"></a>|@wallrj commented|
|-|-|

(In [40406]) Apply udpbroadcast.patch from hawkowl. refs [#11556](https://github.com/twisted/twisted/issues/11556)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_23"></a>|@wallrj commented|
|-|-|

(In [40413]) Apply udpbroadcast-fix.patch from hawkowl. refs [#11556](https://github.com/twisted/twisted/issues/11556)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_24"></a>|@wallrj set owner to @hawkowl|
|-|-|

Hi hawkowl,

This seems to work well. Here are some thoughts:

Points:
 1. When we were discussing this the other day I was assuming that the feature
    was reactor specific, but actually it's only the OS and Python version that
    matters isn't it. So:
    1. Debian: https://buildbot.twistedmatrix.com/builders/debian-easy-py2.6-epoll/builds/1847/steps/epoll/logs/stdio
       twisted.internet.test.test_udp.UDPServerTestsBuilder_SelectReactor.test_allowBroadcast ... [OK]

    2. Fedora17: https://buildbot.twistedmatrix.com/builders/fedora17-x86_64-py2.7/builds/1311/steps/select/logs/stdio
       twisted.internet.test.test_udp.UDPServerTestsBuilder_SelectReactor.test_allowBroadcast ... [OK]

    3. FreeBSD 8.2: https://buildbot.twistedmatrix.com/builders/freebsd-8.2-i386/builds/2222/steps/select/logs/stdio
       twisted.internet.test.test_udp.UDPServerTestsBuilder_SelectReactor.test_allowBroadcast ... [OK]

    4. Ubuntu Natty: https://buildbot.twistedmatrix.com/builders/natty64-py2.6-select/builds/3592/steps/trial/logs/stdio
       twisted.internet.test.test_udp.UDPServerTestsBuilder_SelectReactor.test_allowBroadcast ... [OK]

    5. RHEL6: https://buildbot.twistedmatrix.com/builders/rhel6-x86_64-py2.6/builds/1296/steps/epoll/logs/stdio
       twisted.internet.test.test_udp.UDPServerTestsBuilder_SelectReactor.test_allowBroadcast ... [OK]

    6. Windows7: https://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/3267/steps/select/logs/stdio
       twisted.internet.test.test_udp.UDPServerTestsBuilder_SelectReactor.test_allowBroadcast ... [OK]

    7. Windows XP: https://buildbot.twistedmatrix.com/builders/winxp32-py2.7/builds/2954/steps/select/logs/stdio
       twisted.internet.test.test_udp.UDPServerTestsBuilder_IOCPReactor.test_allowBroadcast ... [OK]

    8. Python3: https://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/1583/steps/shell/logs/stdio
       test_allowBroadcast (twisted.internet.test.test_udp.UDPServerTestsBuilder_SelectReactor)
       test_allowBroadcast ... ok

 2. In that case, it's probably not necessary to add the test to the reactor
    builder test cases.
    * Instead I suggest creating a new static test case class in test_udp

 3. How do you feel about the API? I'm not sure I like it. The implication is
    that we'll have to add separate methods for getting and setting every
    available socket option. (and duplicate the methods for the IOCP
    reactor).

    1. Some other tickets requiring new socket options
       1. SO_REUSEADDR:  [#11362](https://github.com/twisted/twisted/issues/11362)
       2. IPV6_V6ONLY: [#5997](https://github.com/twisted/twisted/issues/5997)
    2. Maybe it would be nicer to define some flag constants which can be passed
       to an endpoint....but that's another ticket...and anyway...

 4. I see that this is how it's been done for various TCP options.

    1. https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.ITCPTransport.html

    2. So we should add setBroadcastAllowed to IUDPTransport?

 5. How about enabling this option automatically if we detect a broadcast
    address or the "<broadcast>" keyword.
    1. https://twistedmatrix.com/trac/browser/tags/releases/twisted-13.1.0/twisted/internet/udp.py#L182

    2. Incidentally, the "<broadcast>" keyword isn't mentioned in the API docs.

 6. And if not, it would be nice to raise something more descriptive than
    "permission denied" if a user attempts to broadcast without having set
    SO_BROADCAST.
    1. [#10933](https://github.com/twisted/twisted/issues/10933)
    2. [#7137](https://github.com/twisted/twisted/issues/7137)

 7. Please add some documentation explaining when UDP broadcasts are useful, how
    they can be enabled and a short example script.

    1. Here: https://twistedmatrix.com/documents/current/core/howto/udp.html

    2. I tested the broadcast feature with the attached script which might be a
       good example. (attachment:udpbroadcast.py)

That's all for now. Please address or answer the points above and resubmit for
review.

Thanks again.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_25"></a>|@hawkowl commented|
|-|-|

Hi rwall,

Thanks for the review! Got some more time to work on this.

1. The static test cases are deprecated - in the reactor builder test cases is the proper place for new tests.
2. I think the API is fine - it's *an* API, rather than breaking encapsulation. The current API matches what's used in the TCP implementation, so it's somewhat consistent, even if it's not perfect. We can deprecate it later anyway.
3. I don't think we should - adding it to IUDPTransport would break other UDP implementations, wouldn't it?
4. Enabling it automatically is non-obvious, I don't think we should.
5. New patch (coming soon!) fixes that.
6. Probably not in the scope of this ticket, I think, but it would be nice.
7. I wrote some long-form docs in the new patch (coming soon). I adapted the multiservice part and general layout from your script there, I hope you don't mind.

The patch will be in a .patch as my internet is currently pretty dodgy. It's based off current trunk.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_26"></a>|@hawkowl commented|
|-|-|

(In [41506]) Branching to udp-broadcast-454-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_27"></a>|@hawkowl commented|
|-|-|

Actually, my internet was good enough!

Looks green so far on the buildbots, so, putting up for review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_29"></a>|@hynek set owner to @hawkowl|
|-|-|

Wow, a 10 years old ticket. :)

Let’s see:

 1. docs/projects/core/examples/udpbroadcast.py
  1. 7: receive
  1. 16: is the `-o` option really necessary? I never used it, does it actually do anything here?
  1. Why a `*` as prompt indicator? I usually use `$ ` but I think something even different was used throughout Twisted.
 1. docs/projects/core/howto/index.rst
  1. I don’t think Multicast and Broadcast should be capitalized.
  1. Is the comma really necessary?
  1. maybe “…including multicast…” is better?
 1. docs/projects/core/howto/udp.rst
  1. 290: I find that sentence unclear, who or what is “it”?
  1. 296: I find it a bit awkward and the “please” superfluous. How about `For a complete example of this feature, see`?
 1. twisted/internet/udp.py
  1. It says “Whether the socket may broadcast” in one case and “Whether this UDP socket may broadcast” in the other; please make this consistent.

Please fix 1.*, 3.1, and 4.1; the rest is up to you. Feel free to merge afterwards if you feel confident enough. :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_30"></a>|@hawkowl set status to `closed`|
|-|-|

(In [41516]) Merge udp-broadcast-454-3: Add UDP broadcast support

Author: hawkowl
Reviewer: exarkun, rwall, hynek
Fixes: [#11556](https://github.com/twisted/twisted/issues/11556)

This adds encapsulation for enabling broadcast on a UDP port.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_31"></a>|GitHub commented|
|-|-|

In changeset 6c6fea1bc2107601a2d270f2ee709ca7e8a24468
```
#!CommitTicketReference repository="" revision="6c6fea1bc2107601a2d270f2ee709ca7e8a24468"
Merge pull request #454 from twisted/8700-einarfd-py3-conch-client-default-3

Author: einarfd, rodrigc
Reviewer: glyph, rodrigc

Port twisted.conch.client.default to Python 3
```
